### PR TITLE
feat: add build sha to parser.c header comment

### PR DESCRIPTION
**Pull request by dundargoc**
- feat: add build sha to parser.c header comment

- fix: make sha generation work with submodules

More specifically, change `is_dir` to `exists` as `.git` is a file when
in a submodule.

### CODE REVIEW

File: cli/generate/src/render.rs [hunk 1]
```diff
@@ -283,9 +283,15 @@ impl Generator {
...
+        const BUILD_SHA: Option<&'static str> = option_env!("BUILD_SHA");
+        let version = BUILD_SHA.map_or_else(
+            || BUILD_VERSION.to_string(),
+            |build_sha| format!("{BUILD_VERSION} ({build_sha})"),
+        );
         add_line!(
             self,
-            "// Automatically generated by tree-sitter v{BUILD_VERSION}"
+            "/* Automatically generated by tree-sitter v{} */",
+            version
```
Comment (amaanq) [inline cli/generate/src/render.rs:294]:
```suggestion
            "/* Automatically generated by tree-sitter v{version} */",```

File: cli/build.rs [hunk 2]
```diff
@@ -6,6 +6,7 @@ use std::{
     time::SystemTime,
 };
 
+#[allow(dead_code)]
```
Comment (amaanq) [inline cli/build.rs:9]: why?

Comment (dundargoc) [reply cli/build.rs:9]: CI fails otherwise.

Comment (amaanq) [reply cli/build.rs:9]: That's because of the build script import - don't import it, just copy paste the function

File: cli/generate/build.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+// This is copied from the build.rs in parent directory. This should be updated if the
+// parent build.rs gets fixes.
+fn read_git_sha() -> Option<String> {
+    let crate_path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap());
+
+    if !crate_path
+        .parent()
+        .unwrap()
```
Comment (ObserverOfTime) [inline cli/generate/build.rs:16]: This will panic if there is no parent.

File: cli/generate/src/render.rs [hunk 4]
```diff
@@ -283,9 +283,14 @@ impl Generator {
     }
 
     fn add_header(&mut self) {
+        const BUILD_SHA: Option<&'static str> = option_env!("BUILD_SHA");
```
Comment (ObserverOfTime) [inline cli/generate/src/render.rs:286]: Move this to the top level.

Comment (ObserverOfTime) [reply cli/generate/src/render.rs]: Micro-optimisation.

File: cli/generate/build.rs (hunk 3 above)
Comment (dundargoc) [reply cli/generate/build.rs:17]: Added a `Some(crate_path.parent())?;` check before to return early. Should work, but unsure if it's idioamtic rust or whatever.

File: cli/generate/build.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+// This is copied from the build.rs in parent directory. This should be updated if the
+// parent build.rs gets fixes.
+fn read_git_sha() -> Option<String> {
+    let crate_path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap());
+    Some(crate_path.parent())?;
+
+    if !crate_path
+        .parent()
+        .unwrap()
```
Comment (amaanq) [inline cli/generate/build.rs:17]:
```suggestion
        .parent()?
```

File: cli/generate/build.rs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    if let Some(git_sha) = read_git_sha() {
+        println!("cargo:rustc-env=BUILD_SHA={git_sha}");
+    }
+}
+
+// This is copied from the build.rs in parent directory. This should be updated if the
+// parent build.rs gets fixes.
+fn read_git_sha() -> Option<String> {
+    let crate_path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap());
+    Some(crate_path.parent())?;
```
Comment (amaanq) [inline cli/generate/build.rs:13]:
```suggestion
```

**Review by ObserverOfTime — APPROVED** [review]

**Pull request merged**